### PR TITLE
Export fi_var_* functions as 1.0 API.

### PR DESCRIPTION
**Pull request by pmmccorm**
This fixes dl-provider builds.

Also add fi_getvars(struct env **env, int *count) which
will allocate and return all current provider registered
environment variables.

The caller can free the env with free()

Signed-off-by: pmmccorm <patrick.m.mccormick@intel.com>

### CODE REVIEW

File: src/var.c [hunk 1]
```diff
@@ -53,8 +55,37 @@ struct fi_var_t {
...
 static struct fi_var_t *fi_vars = NULL;
 
+__attribute__((visibility ("default")))
+int DEFAULT_SYMVER_PRE(fi_getvars)(struct fi_env **vars, int *count) {
+
+	struct fi_env *vhead;
+	int len = 0;
+
+	// just get a count
+	for (struct fi_var_t *ptr = fi_vars; ptr ;ptr = ptr->next, ++len) {}
```
Comment (shefty) [inline src/var.c]: nitpicking:
Move {} to next line (and consider replacing with just ;
Just declare ptr variable at the top of the call, rather than declaring it twice
Also, leave a space after the ; but not before in the for loop

File: src/var.c [hunk 2]
```diff
@@ -53,8 +55,37 @@ struct fi_var_t {
...
+
+	vhead = calloc(len, sizeof (*vhead));
+
+	if (!vhead)
+		return -FI_ENOMEM;
+
+	int i = 0;
```
Comment (shefty) [inline src/var.c]: replace len and i with just i - please don't declare variables in the middle of functions, it makes locating their declaration difficult

File: include/rdma/fi_var.h [hunk 3]
```diff
@@ -40,6 +40,13 @@
 extern "C" {
 #endif
 
+struct fi_env {
+	const struct fi_provider *provider;
```
Comment (shefty) [inline include/rdma/fi_var.h]: This is giving apps a hook directly into the provider calls.  I don't think we want to support this.  Just provide the provider name.

File: src/var.c [hunk 4]
```diff
@@ -53,8 +55,37 @@ struct fi_var_t {
...
+	for (struct fi_var_t *ptr = fi_vars; ptr; ptr = ptr->next, ++i) {
+		vhead[i] = (struct fi_env) {	.provider = ptr->provider,
+						.var_name = ptr->var_name,
+						.env_var_name = ptr->env_var_name,
+						.help_string = ptr->help_string
+					   };
```
Comment (shefty) [inline src/var.c]: Is there a reason why the code is structure like this, versus just doing:

vars[i].var_name = ptr->var_name;
vars[i].env_var_name = ptr->env_var_name;
...

Note that this is still a racy API.  This returns pointers to internal structures, which means that the library can never free this memory after this call is made.  Use strdup() instead.  And add an fi_freevars() call

File: libfabric.map [hunk 5]
```diff
@@ -9,5 +9,11 @@ FABRIC_1.0 {
 		fi_tostr;
 		fi_log_enabled;
 		fi_log;
+		fi_var_register;
+		fi_var_get_str;
+		fi_var_get_int;
+		fi_var_get_long;
+		fi_var_get_bool;
```
Comment (shefty) [inline libfabric.map]: These changes should be in a separate commit.

File: include/rdma/fi_var.h (hunk 3 above)
Comment (pmmccorm) [inline include/rdma/fi_var.h]: I included the fi_provider pointer so we could pass it to the various fi_var_get_* functions.

File: src/var.c (hunk 1 above)
Comment (pmmccorm) [inline src/var.c]: Will do.

File: src/var.c (hunk 2 above)
Comment (pmmccorm) [inline src/var.c]: OK.

File: src/var.c (hunk 4 above)
Comment (pmmccorm) [inline src/var.c]: I could do either style for assignment, I am not tied to either one.

Racy API: yes two threads can have an inconsistent view of the environment vars, but adding locking won't change this right?

This returns pointers to internal structures but the only memory alloc'd is the storage to store the pointers, so:

```
fi_free_vars(struct fi_env *env) { 
    free(env);
}
```

Won't free any internal vars and is all we need right?

File: src/var.c (hunk 1 above)
Comment (jsquyres) [inline src/var.c]: Another nit pick: maybe use "continue" instead of a blank ";"...?

**Comment by pmmccorm** [comment]
Rebased and updated.

File: src/var.c [hunk 6]
```diff
@@ -53,8 +56,86 @@ struct fi_var_t {
...
+__attribute__((visibility ("default")))
+int DEFAULT_SYMVER_PRE(fi_getsettings)(struct fi_setting **vars, int *count) {
+
+	struct fi_setting *vhead = NULL;
+	struct fi_var_t *ptr;
+	int ret, len = 0, i = 0;
+	char *tmp;
+
+	// just get a count
+	for (ptr = fi_vars; ptr; ptr = ptr->next, ++len);
```
Comment (jsquyres) [inline src/var.c]: Might have gotten lost because I commented right before you re-pushed, but this format is nice IMHO:

```c
    for (ptr = fi_vars; ptr; ptr = ptr->next, ++len)
        continue;
```

You still see an obvious block under the for line, etc.

But that's just my $0.02.  :smile:

Comment (pmmccorm) [inline src/var.c]: Yeah that is more clear. I updated and pushed.

File: src/var.c [hunk 7]
```diff
@@ -53,8 +56,87 @@ struct fi_var_t {
...
+		vhead[i].var_name = strdup(ptr->var_name);
+		vhead[i].env_var_name = strdup(ptr->env_var_name);
+		vhead[i].help_string = strdup(ptr->help_string);
 
-int fi_var_register(const struct fi_provider *provider, const char *var_name,
+		/* ignore error for now */
+		ret = fi_var_get(ptr->provider, ptr->var_name, &tmp);
+
+		if (ret == FI_SUCCESS)
+			vhead[i].value = strdup(tmp);
```
Comment (shefty) [inline src/var.c]: Add checks against strdup failing, so that coverity doesn't immediately complain after this is merged.

File: src/var.c [hunk 8]
```diff
@@ -53,8 +56,87 @@ struct fi_var_t {
...
+
+void DEFAULT_SYMVER_PRE(fi_freesettings)(struct fi_setting *var) {
+	if (var) {
+		free((void*)var->prov_name);
+		free((void*)var->var_name);
+		free((void*)var->env_var_name);
+		free((void*)var->help_string);
+		free((void*)var->value);
+		free (var);
+	}
```
Comment (shefty) [inline src/var.c]: This needs some sort of loop.  Maybe add an empty fi_setting struct to the end of the array.

**Comment by jsquyres** [comment]
@pmmccorm I'm trying to reproduce the failure you're seeing so that I can add it to our Jenkins runs.  ...but I'm unable to reproduce it with the current libfabric master.  What am I doing wrong?

```sh
# In libfabric
$ ./configure --enable-usnic=dl --enable-sockets=dl --enable-verbs=dl --prefix=blah
$ make -j32 clean install

# Over in fabtests
$ ./configure --with-libfabric=blah --prefix=blah
$ make -j32 clean all
```

It doesn't fail to build, and I can run the tests (e.g., `simple/fi_info`).

What am I missing?

**Comment by pmmccorm** [comment]
That should result in an error. Does your built libfabric have exported fi_var_* symbols ? If so the =dl build didn't work.

File: src/var.c (hunk 8 above)
Comment (pmmccorm) [inline src/var.c]: Der yeah, I'll fix that.

**Comment by jsquyres** [comment]
Ok, I got it now.  It was a PEBKAC.  Now I can replicate.

**Comment by pmmccorm** [comment]
Updated to check strdup's for failure, properly free settings structures.

**Comment by jsquyres** [comment]
Our internal Jenkins is now checking dl builds.  So the sooner this gets merged, the sooner our Jenkins will start passing again.  :-)

File: src/var.c [hunk 9]
```diff
@@ -53,8 +56,98 @@ struct fi_var_t {
...
+	for (ptr = fi_vars; ptr; ptr = ptr->next, ++len)
+		continue;
+
+	if (len == 0)
+		goto no_vars;
+
+	// last extra entry will be all NULL
+	vhead = calloc(len+1, sizeof (*vhead));
+
+	if (!vhead)
```
Comment (shefty) [inline src/var.c]: nit: remove blank line between setting vhead and checking result

File: src/var.c [hunk 10]
```diff
@@ -53,8 +56,98 @@ struct fi_var_t {
...
+	for (ptr = fi_vars; ptr; ptr = ptr->next, ++i, tmp = NULL) {
+		vhead[i].prov_name = strdup(ptr->provider->name);
+		vhead[i].var_name = strdup(ptr->var_name);
+		vhead[i].env_var_name = strdup(ptr->env_var_name);
+		vhead[i].help_string = strdup(ptr->help_string);
+
+		// ignore error for now, maybe skip in the future?
+		ret = fi_var_get(ptr->provider, ptr->var_name, &tmp);
+
+		if (ret == FI_SUCCESS && tmp)
```
Comment (shefty) [inline src/var.c]: nit: remove blank line before if check.
Remove comment above call to fi_var_get, return value is checked.

**Comment by shefty** [comment]
I think this is fine to merge now.  We need to add man page entries for the new calls.
